tests/two_d_array_int/two_d_array_int.c

```diff
@@ -0,0 +1,251 @@
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        TWO_D_ARRAY_FREE_ROW(TEST_THANDLE)(tdarr, i);
+    }
+
+    //assert
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        ASSERT_IS_NULL(tdarr->row_arrays[i]);
+    }
+
+    //cleanup
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_allocate_a_row_already_been_allocated)
+{
+    //arrange
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(ROW_NUM, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+    int result;
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, 0);
+    ASSERT_ARE_EQUAL(int, result, 0);
+
+    //act
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, 0);
+
+    //assert
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+
+    //cleanup
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_GET_ROW_succeeds)
+{
+    //arrange
+    TEST_THANDLE* row_result;
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, 0);
+
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    for (unsigned int i = 0; i < 5; i++)
+    {
+        tdarr->row_arrays[0][i].a = i;
+    }
+
+    //act
+    row_result = TWO_D_ARRAY_GET_ROW(TEST_THANDLE)(tdarr, 0);
+
+    //assert
+    ASSERT_IS_NOT_NULL(row_result);
+    for (unsigned int i = 0; i < 5; i++)
+    {
+        ASSERT_ARE_EQUAL(int64_t, row_result[i].a, i);
+    }
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_get_a_row_not_allocated_yet)
+{
+    //arrange
+    TEST_THANDLE* row_result;
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(5, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+
+    //act
+    row_result = TWO_D_ARRAY_GET_ROW(TEST_THANDLE)(tdarr, 0);
+
+    //assert
+    ASSERT_IS_NULL(row_result);
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_free_a_row_not_allocated_yet)
+{
+    //arrange
+    int result;
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(5, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+    for (uint32_t i = 0; i < 5; i++)
+    {
+        ASSERT_IS_NULL(tdarr->row_arrays[i]);
+    }
+
+    //act
+    result = TWO_D_ARRAY_FREE_ROW(TEST_THANDLE)(tdarr, 0);
+
+    //assert
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+#define N_CHAOS_API_CALLS 5000
+TEST_FUNCTION(TWO_D_ARRAY_chaos)
+{
+    //arrange
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(ROW_NUM, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, i);
+    }
+
+    //action
+    uint32_t run_time = 0;
+    while (run_time < N_CHAOS_API_CALLS)
+    {
+        run_time++;
```

Comment:
>run_time [](http://example.com/codeflow?start=8&length=8)

seems like this is `iteration_count` as `run_time` implies we are doing some time based test.

Reply:
modified